components/cereal-service/pkg/server/server.go

```diff
@@ -0,0 +1,632 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/sirupsen/logrus"
+	"github.com/teambition/rrule-go"
+
+	"github.com/chef/automate/api/interservice/cereal"
+	libcereal "github.com/chef/automate/lib/cereal"
+	"github.com/chef/automate/lib/cereal/backend"
+)
+
+var errInvalidMsg = errors.New("invalid msg")
+
+type CerealService struct {
+	workflowScheduler *libcereal.WorkflowScheduler
+	backend           backend.Driver
+}
+
+func NewCerealService(ctx context.Context, b backend.Driver) *CerealService {
+	cs := &CerealService{
+		backend: b,
+	}
+
+	if v, ok := b.(backend.SchedulerDriver); ok {
+		cs.workflowScheduler = libcereal.NewWorkflowScheduler(v)
+		go cs.workflowScheduler.Run(ctx)
+	}
+
+	return cs
+}
+
+func generateRequestID() string {
+	id, err := uuid.NewV4()
+	if err != nil {
+		return "unknown"
+	}
+	return id.String()
+}
+
+func (s *CerealService) EnqueueWorkflow(ctx context.Context, req *cereal.EnqueueWorkflowRequest) (*cereal.EnqueueWorkflowResponse, error) {
+	logctx := logrus.WithFields(logrus.Fields{
+		"id":            generateRequestID(),
+		"method":        "EnqueueWorkflow",
+		"workflow_name": req.WorkflowName,
+		"instance_name": req.InstanceName,
+	})
+	logctx.Info("enqueuing workflow")
+	if err := s.backend.EnqueueWorkflow(ctx, &backend.WorkflowInstance{
+		InstanceName: req.InstanceName,
+		WorkflowName: req.WorkflowName,
+		Parameters:   req.Parameters,
+	}); err != nil {
+		logctx.WithError(err).Error("failed to enqueue workflow")
+		if err == libcereal.ErrWorkflowInstanceExists {
+			return nil, status.Error(codes.FailedPrecondition, err.Error())
+		}
+	}
+	return &cereal.EnqueueWorkflowResponse{}, nil
+}
+
+func readDeqWorkReqMsg(ctx context.Context, s cereal.Cereal_DequeueWorkflowServer) (*cereal.DequeueWorkflowRequest, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	type chanMsg struct {
+		Msg *cereal.DequeueWorkflowRequest
+		Err error
+	}
+
+	in := make(chan chanMsg)
+	go func() {
```

Comment:
I'm not 100% following why these goroutines are necessary.

Reply:
the send and recv functions don't take a context. Their underlying context expires when you return from the function that is part of the grpc interface that was generated. If you want to do a read with a context (for example one that times out), you need to do it in another goroutine. it's horrible

Reply:
Ah, I missed the fact that the calling function is passing a context that has been given a timeout so I was confused about why we were going through all of this trouble when the context would only be canceled by the client going away anyway, but now that I see the Timeout it makes a lot more sense.